### PR TITLE
feat: add Hugging Face LoRA fine-tuning job bundle

### PR DESCRIPTION
**Pull request by anika-suman-amazon**
### What was the problem/requirement? (What/Why)

`deadline-cloud-samples` had bundles for image LoRA training (`flux2_klein_lora`) and LLM evaluation (`vllm_lm_eval_leaderboard`) but nothing for **text LLM fine-tuning** — one of the most common customer requests (teaching a base model brand voice, product knowledge, or output formats). Without a sample, customers had to build the whole orchestration themselves.

### What was the solution? (How)

A new bundle `job_bundles/hf_finetune_lora/` that fine-tunes any HuggingFace causal LM with LoRA / QLoRA on a custom instruction dataset:

- **`template.yaml`** + **`train.py`** — QLoRA training via `transformers`, `peft`, `bitsandbytes`, `datasets`
- **Dataset flexibility** — accepts a **local folder** (via job attachments) or **S3 URI** (single file or prefix); recursively concatenates all `.jsonl` files
- **`sample_data/`** — a bundled fictional-restaurant-chain dataset split across 5 thematic files so submitting with all defaults produces a working demo out-of-the-box
- **`examples/saffron_stack/`** — canonical lore doc + generator showing how to build your own domain dataset
- **`inference/`** — client-side `chat.py` REPL and `gradio_chat.py` web UI with live base-vs-tuned toggle and automatic output cleanup

Default hyperparameters are tuned for fact memorization — the harder end of fine-tuning. Lighter configs for style-transfer are documented in the README.

### What is the impact of this change?

- **Completes the LLM lifecycle story** in the samples repo: train (this) → evaluate (`vllm_lm_eval_leaderboard`) → serve (external--bundle outputs standard PEFT format)
- **Zero-to-working-chatbot in one command**: submitting with all defaults yields a working fine-tuned adapter using the bundled sample data

### How was this change tested?

Tested end-to-end on a Deadline Cloud farm with a Linux fleet + NVIDIA L4 GPU. Verified:

- **Smoke test** — smallest possible config (Qwen 0.5B, tiny sample count, 1 epoch) submits, runs, and produces an adapter
- **Full training run** with the default hyperparameters — loss decreases monotonically, adapter saves to `OutputDir/AdapterName/`, downloads correctly via `deadline job download-output`
- **Built-in `--run-tests` suite** — after training, the script runs a set of fact-recall questions against the just-trained adapter and prints answers to the worker log. Verified the fine-tuned model produces adapter-specific answers that the base model cannot produce (proving the fine-tuning took effect)
- **Multi-file dataset loading** — recursive `.jsonl` glob works on both flat (single folder) and nested (subfolder) layouts
- **Both dataset sources** — `DatasetPath` (local folder auto-uploaded via job attachments) and `DatasetS3Uri` (S3 single file *and* S3 prefix) both work end-to-end
- **GUI submitter** — `deadline bundle gui-submit` renders all parameter groups correctly with pre-populated conda queue-env values
- **Client-side tools** — `chat.py` REPL and `gradio_chat.py` web UI both load the adapter, run inference, and provide the base-vs-tuned toggle

### Was this change documented?

Yes:

- `README.md` — full usage, parameters, IAM setup, examples
- `sample_data/README.md`, `examples/saffron_stack/README.md`, `examples/saffron_stack/LORE.md`, `inference/README.md` — per-directory docs
- Updated `job_bundles/README.md` (top-level samples index) with a new entry immediately after the existing LLM evaluation sample

---

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: job_bundles/hf_finetune_lora/train.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+def main() -> int:
+    args = parse_args()
+
+    # Print environment summary up front to make debugging easy.
+    log(f"PyTorch: {torch.__version__}, CUDA available: {torch.cuda.is_available()}")
+    if torch.cuda.is_available():
+        log(f"GPU: {torch.cuda.get_device_name(0)}, "
+            f"VRAM: {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")
+    log(f"Args: {vars(args)}")
```
[the remark this answers is not in the document]
Comment (Cherie-Chen) [reply job_bundles/hf_finetune_lora/train.py]: We are following the pattern from other sample to pass in HF Token in plain text, this is acceptable. But, please do not log it. If you want to know if the token is passed in, do what was recommended by the bot

File: job_bundles/hf_finetune_lora/template.yaml [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    label: Base Model
+    groupLabel: Model
+  default: "Qwen/Qwen2.5-7B"
+  allowedValues:
+  - "Qwen/Qwen2.5-1.5B"
+  - "Qwen/Qwen2.5-7B"
+  - "Qwen/Qwen2.5-0.5B"
+  - "meta-llama/Llama-3.1-8B"
+  - "mistralai/Mistral-7B-v0.3"
+  - "microsoft/Phi-3.5-mini-instruct"
```
Comment (Cherie-Chen) [inline job_bundles/hf_finetune_lora/template.yaml:20]: Have we tested all of these? 

Base models for `meta-llama/Llama-3.1-8B` and `mistralai/Mistral-7B-v0.3` does not include chat_template and may not work?

File: job_bundles/hf_finetune_lora/train.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    # Prefix — list and download all *.jsonl objects under it
+    prefix = key if key.endswith("/") else key + "/"
+    paginator = s3.get_paginator("list_objects_v2")
+    downloaded = []
+    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
+        for obj in page.get("Contents", []):
+            if not obj["Key"].endswith(".jsonl"):
+                continue
+            fname = os.path.basename(obj["Key"])
+            local_path = os.path.join(local_dir, fname)
```
[the remark this answers is not in the document]
Comment (Cherie-Chen) [reply job_bundles/hf_finetune_lora/train.py]: This is a valid concern. 

`list_objects_v2` returns keys recursively, so two objects under different sub-prefixes with the same filename (e.g. `data/lunch/menu.jsonl` and `data/dinner/menu.jsonl`) map to the same `local_path`. The second download overwrites the first, and the same path gets appended to downloaded twice — so one file's records are dropped entirely while the survivor is loaded twice. No error fires (the empty-list guard doesn't catch it), so training silently runs on corrupted data.

File: job_bundles/hf_finetune_lora/inference/gradio_chat.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    tokenizer = AutoTokenizer.from_pretrained(base_model_id)
+    base = AutoModelForCausalLM.from_pretrained(base_model_id, torch_dtype=dtype).to(device)
+    base.eval()
+    model = PeftModel.from_pretrained(base, str(adapter_path))
+    model.eval()
+    print("Ready. Launching web UI...")
+
+    def chat_fn(message: str, history: list[list[str]],
+                use_adapter: bool, max_new_tokens: int) -> str:
```
Comment (Cherie-Chen) [inline job_bundles/hf_finetune_lora/inference/gradio_chat.py]: NIT: The annotation `list[list[str]]` for `history` is the deprecated Gradio "tuples" shape (removed in Gradio 6.0 in favor of `list[dict]` messages in https://github.com/gradio-app/gradio/pull/12012). Since the value is never used it has no runtime effect — and with `from __future__ import annotations` at the top it isn't even evaluated — but if you're targeting 6.0 syntax it's worth updating to `list[dict]` (or dropping the annotation) so it doesn't mislead anyone who later wants to add multi-turn support.

Suggested:
```python
def chat_fn(message: str, _history: list[dict],
            use_adapter: bool, max_new_tokens: int) -> str:
```

**Comment by Cherie-Chen** [comment]
Please remove `.crux_template.md` file

File: job_bundles/hf_finetune_lora/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anika-suman-amazon) [reply job_bundles/hf_finetune_lora/train.py]: Done, redacted hf_token to *** before logging args!

File: job_bundles/hf_finetune_lora/inference/gradio_chat.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    p.add_argument("--adapter-path", required=True,
+                   help="Path to the LoRA adapter directory")
+    p.add_argument("--base-model", default=None,
+                   help="HF base model ID. If omitted, read from training_metadata.json.")
+    p.add_argument("--max-new-tokens", type=int, default=100,
+                   help="Default max tokens per response. You can also adjust this "
+                        "live via the slider in the UI.")
+    p.add_argument("--temperature", type=float, default=0.0,
+                   help="0 = greedy (deterministic), >0 = sampled")
+    p.add_argument("--share", action="store_true",
```
[the remark this answers is not in the document]
Comment (anika-suman-amazon) [reply job_bundles/hf_finetune_lora/inference/gradio_chat.py:49]: Added warnings that the tunnel is public and unauthenticated to the --share help text, module docstring, and inference README. Kept it warning-only (no auth) since this is a local dev tool run on the user's own machine — the exposure is now explicit and opt-in.

File: job_bundles/hf_finetune_lora/inference/gradio_chat.py (hunk 4 above)
Comment (anika-suman-amazon) [reply job_bundles/hf_finetune_lora/inference/gradio_chat.py]: Updated to list[dict] and renamed the unused param to _history.

File: job_bundles/hf_finetune_lora/template.yaml (hunk 2 above)
Comment (anika-suman-amazon) [reply job_bundles/hf_finetune_lora/template.yaml:20]: Removed Llama-3.1-8B (gated:manual, needs an approved access request) from the dropdown. Added a ChatML fallback chat_template so base checkpoints without one (e.g. Mistral-7B-v0.3) train instead of crashing. Also fixed the stale "Mistral is gated" note because it turns out that that specific model is public! I asked Claude to select a few models so the ones in the dropdown are now all ungated. I haven't tested all of them but I tested a couple!

File: job_bundles/hf_finetune_lora/train.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anika-suman-amazon) [reply job_bundles/hf_finetune_lora/train.py]: Fixed, downloads now preserve each key's path relative to the prefix (recreating subdirs), so same-named files under different sub-prefixes no longer collide. Matches the local rglob behavior.

**Review by Cherie-Chen — APPROVED** [review]
